plugins/com.google.cloud.tools.eclipse.googleapis/src/com/google/cloud/tools/eclipse/googleapis/internal/DefaultAccountProvider.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.eclipse.googleapis.internal;
+
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
+import com.google.api.client.googleapis.util.Utils;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpRequestInitializer;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.services.oauth2.Oauth2;
+import com.google.cloud.tools.eclipse.googleapis.Account;
+import com.google.cloud.tools.eclipse.googleapis.UserInfo;
+import com.google.cloud.tools.eclipse.util.CloudToolsInfo;
+import java.io.IOException;
+import java.util.logging.Logger;
+
+/**
+ * 
+ */
+public class DefaultAccountProvider implements IAccountProvider {
+
+  private static final int USER_INFO_QUERY_HTTP_CONNECTION_TIMEOUT = 5000 /* ms */;
+  private static final int USER_INFO_QUERY_HTTP_READ_TIMEOUT = 3000 /* ms */;
+  private static final HttpTransport transport = new NetHttpTransport();
+  
+  private final JsonFactory jsonFactory = Utils.getDefaultJsonFactory();
+  
+  private static final HttpRequestInitializer requestTimeoutSetter = new HttpRequestInitializer() {
+    @Override
+    public void initialize(HttpRequest httpRequest) throws IOException {
+      httpRequest.setConnectTimeout(USER_INFO_QUERY_HTTP_CONNECTION_TIMEOUT);
+      httpRequest.setReadTimeout(USER_INFO_QUERY_HTTP_READ_TIMEOUT);
+    }
+  };
+  
+  /**
+   * @return the application default credentials associated account
+   */
+  @Override
+  public Account getAccount() throws IOException{
+     return getAccount(GoogleCredential.getApplicationDefault());
+  }
+  
+  /**
+   * @return the ADC if set
+   */
+  @Override
+  public Credential getCredential() throws IOException {
+    if (hasCredentialsSet()) {
+      return null;
+    }
+    return getAccount().getOAuth2Credential();
+  }
+  
+  /**
+   * Convenience method to determine if the user has ADC set
+   * @return true if the gcloud CLI has Application Default Credentials set
+   */
+  @Override
+  public boolean hasCredentialsSet() throws IOException {
+    return getAccount() != null && getAccount().getOAuth2Credential() != null;  
```

Comment:
> @diegomarquezp : [Paraphrased] Should we throw these `IOException`s when unable to get the credential?

Generally, it depends on whether the `IOException` simply represents "no credential", or whether it is truly "exceptional" in the sense that "something unexpected went wrong". Also, generally, it depends on whether the caller of the function would potentially be capable of handling / recovering from the exception in some way that the current method cannot.

I would generally expect a boolean-returning method like `hasCredentialsSet` to return false if the `IOException` is simply `getOAuth2Credential()`'s standard behavior for saying "no credential".

>  @diegomarquezp : [Paraphrased] Should we return `null` from `getCredential()`?

I've jumped on the Java 8 train, and returning `null` is against my religion. Instead, I would be returning an never-null `Optional<Credential>` reference. However, the rest of this code base tends to return `null` often, so there is an argument to be made that maintaining the existing standard is okay.

Ultimately these are trade-offs that depends on the context and how heavily you weigh each consideration. Given the above, what do you think is the best choice in this situation that lets you achieve the desired goal?

Reply:
In that sense, the `IOException` occurs when something goes wrong when obtaining the user info via an HTTP request using the default credentials as base. It is therefore truly exceptional. I will modify this to bubble up the exception.

Reply:
Actually, I think it's fine to bubble it up to `GoogleApiFactory` where then the exception is catched, but there is still a null check for the credentials in every `new[...]Api()` method. I think it's fair to leave external callers unconcerned with a failed internal HTTP request to obtain user info.